lessons/python_basics/practice_code.py

```diff
@@ -0,0 +1,277 @@
+#Here is a smattering of commands.
+
+s = "This is a test string"
+print s
+print s[0]			#Array syntax in python - to access element i of an array, we do array[i]
+print s[-3]			#Negative indicies access elements backwards from the end of the list
+print s[2:4]		#Accesses elements 2 - 3
+print s[6:]			#Accesses elements 6 and above
+print s + ' ' + s	#Strings can be concatenated
+print s*10			#Or multiplied...
+
+print s.replace('a test','the bestest') # print a string with a replaced word
+
+#Above, replace is a method, a function that is a part of an 
+#object (in this case, s).  Tab-completion can be used to determine 
+#what methods exist for a given object.
+
+#Perform equality tests on strings; the output is a boolean.
+#(Python doesn't care about using ' or " for strings.)
+
+print s == 'This is a test string'
+
+
+
+########## DATA TYPES: LISTS, TUPLES & DICTS ##########
+
+#Python has two array-like things.  The first is called a "list", 
+#which can hold any data types:
+
+mylist = [1, "2", 3, 4.000, "five"]
+print len(mylist)						#Prints number of elements in list
+print mylist[2:]						#Indexing is the same as for strings
+print mylist + mylist					#Adding lists combines them
+print mylist.index(4.)					#Return the index of a list element
+
+#The second is called a "tuple", which is an immutable list (nothing 
+#can be added or subtracted) whose elements also can't be reassigned.  
+#They're easier on memory and processor than lists.
+
+mytuple = (1, "2", 3j, 4.000, "five")
+
+#Python also has dictionaries, which are unordered sets of values accessible 
+#by string keys.
+
+mydict = {"dogs": "DOGS!", "dognumber": 47}			#Can also just write {} for a blank dictionary
+mydict["dognames"] = ["spot", "rover", "agamemnon"]	#Can add dictionary entries after initialization
+print mydict										#Print is versatile, and will give you a summary of your dict
+del mydict["dogs"]									#Delete a key and its associated value
+
+#Note that passing an "immutable type" of object, like a float, tuple 
+#or int, using "=" will make a copy of the value.
+
+x = 4
+y = x
+print x, y
+y = 7
+print x, y
+
+#Lists and dicts, however, are "mutable", and 
+
+x = ["alpha", "beta"]
+y = x
+print x, y
+y[1] = "gamma"
+print x, y
+
+#To make a copy of a list, do:
+
+y = list(x)		#x[:] also works
+y[1] = "delta"
+print x, y
+
+#See https://en.wikibooks.org/wiki/Python_Programming/Data_Types#Mutable_vs_Immutable_Objects
+
+
+########## IMPORTING LIBRARIES ##########
+
+#If you want Matlab-style arrays, you'll need to install the 
+#numpy library.  Here, "try/except" means try <try code>, and 
+#if Python throws up an exception, do <except code>.
+
+try:
+    import numpy as np						#Imports numpy library
+    myarray = np.array([1,43,32.1,17,216])	#Create numpy array
+    print myarray
+    print myarray*myarray/3.				#Multiplication is element by element; "3." is typecast to an array of length myarray
+except:
```

Comment:
The `try` should contain the least amount of code as possible and the `except` should be as specific as possible. Consider if you misspelled `myarray`; that would be a `NameError`, but you'd never see it because of the `try`/`except`. You can make this work using `else`:

```python
try:
    import numpy as np						#Imports numpy library
except ImportError:
    print "You don't have numpy installed!  This section of code won't run." 
else:
    myarray = np.array([1,43,32.1,17,216])	#Create numpy array
    print myarray
    print myarray*myarray/3.				#Multiplication is element by element; "3." is typecast to an array of length myarray
```